libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/factory/RoadObjectFactory.kt

```diff
@@ -71,6 +65,17 @@ object RoadObjectFactory {
         roadObject: RoadObject,
         distanceToStart: Double?,
         distanceInfo: RoadObjectDistanceInfo?
-    ) =
-        UpcomingRoadObject(roadObject, distanceToStart, distanceInfo)
+    ) = UpcomingRoadObject(
+        roadObjectProvider = { roadObject },
+        distanceToStartProvider = { distanceToStart },
+        distanceInfoProvider = { distanceInfo }
```

Comment:
This method is to allow the creation of the `UpcomingRoadObject` if you already have the SDK RoadObject.

Can we refactor it? We could, but I don't think we need to. As you noticed, It's only called by the `RoadObjectsStore`, and it's called after [native RoadObject] has already been mapped to the [SDK RoadObject]. 

Refactoring it would require changing its signature to accept `UpcomingRoadObject` property initializer functions:

```kotlin
// RoadObjectFactory
fun buildUpcomingRoadObject(
    roadObjectProvider: () -> RoadObject,
    distanceToStartProvider: () -> Double?,
    distanceInfoProvider: () -> RoadObjectDistanceInfo?,
) = UpcomingRoadObject(
    roadObjectProvider = roadObjectProvider,
    distanceToStartProvider = distanceToStartProvider,
    distanceInfoProvider = distanceInfoProvider
)
```
and then modifying how `RoadObjectsStore` creates the list of `UpcomingRoadObjects`, to inject property initializer functions for each constructed `UpcomingRoadObject`:

```kotlin
// RoadObjectsStore
fun getUpcomingRoadObjects(
    distances: List<RoadObjectDistanceInfo>
): List<UpcomingRoadObject> {
    val upcomingObjects = mutableListOf<UpcomingRoadObject>()
    distances.forEach {
        navigator.roadObjectsStore?.getRoadObject(it.roadObjectId)?.also { nativeRoadObject ->
            upcomingObjects.add(
                RoadObjectFactory.buildUpcomingRoadObject(
                    roadObjectProvider = { RoadObjectFactory.buildRoadObject(nativeRoadObject) },
                    distanceToStartProvider = { it.distanceToStart },
                    distanceInfoProvider = { it }
                )
            )
        }
    }

    return upcomingObjects
}
```

WDYT?

Reply:
That's an option but ultimately I'd like us to go a level deeper and lazy initialize all road objects. Now we'd have to do it for each mapping function (for road object store, for upcoming road object, for road object matchers, etc) but what if we did something like this instead:
```diff
diff --git a/libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/roadobject/RoadObject.kt b/libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/roadobject/RoadObject.kt
index 9da45c8a5a..4a7ec815ae 100644
--- a/libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/roadobject/RoadObject.kt
+++ b/libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/roadobject/RoadObject.kt
@@ -1,5 +1,6 @@
 package com.mapbox.navigation.base.trip.model.roadobject
 
+import com.mapbox.navigation.base.trip.model.eh.mapToRoadObjectLocation
 import com.mapbox.navigation.base.trip.model.roadobject.location.RoadObjectLocation
 import com.mapbox.navigation.base.trip.model.roadobject.location.RoadObjectLocationType
 
@@ -43,12 +44,13 @@ abstract class RoadObject internal constructor(
     val id: String,
     val objectType: Int,
     val length: Double?,
-    val location: RoadObjectLocation,
     val provider: String,
     val isUrban: Boolean?,
     internal val nativeRoadObject: com.mapbox.navigator.RoadObject,
 ) {
 
+    val location: RoadObjectLocation by lazy { nativeRoadObject.location.mapToRoadObjectLocation() }
+
     /**
      * Indicates whether some other object is "equal to" this one.
      */
```

This way regardless of who's the generator of the object, the heavy mapping is always lazy. We could do it just for `location`, or any other heavy property.

Reply:
I like that idea. Let me try this and run some performance analysis.